hack/verify-upgrade.sh

```diff
@@ -0,0 +1,188 @@
+
+echo "Installing cert-manager ${LATEST_RELEASE} Helm chart into the cluster..."
+
+# Upgrade or install latest published cert-manager Helm release
+helm upgrade \
+    --install \
+    --wait \
+    --namespace "${NAMESPACE}" \
+    --set installCRDs=true \
+    --create-namespace \
+    --version "${LATEST_RELEASE}" \
+    "$RELEASE_NAME" \
+    "$HELM_CHART"
+
+# TODO: This is a dumb sleep that must be removed once we have a check for webhook to be ready.
+echo "Sleep for 30s whilst waiting for cert-manager installation to be ready..."
+sleep 30
+
+# Create a cert-manager issuer and cert
+kubectl apply -f "${REPO_ROOT}/test/fixtures/cert-manager-resources.yaml" \
+	--selector=test=first
+
+# Ensure cert becomes ready
+ensure_cert_ready "test1"
+
+# 1. BUILD AND UPGRADE TO HELM CHART FROM THE CURRENT MASTER
+
+echo "Upgrading cert-manager Helm release to ${CURRENT_VERSION}..."
+"${REPO_ROOT}/devel/addon/certmanager/install.sh"
+
+# Test that the existing cert-manager resources can still be retrieved
+kubectl get issuer/selfsigned-issuer cert/test1
+
+# Create another certificate
+kubectl apply -f "${REPO_ROOT}/test/fixtures/cert-manager-resources.yaml" \
+	--selector=test=second
+
+# Ensure cert becomes ready
+ensure_cert_ready "test2"
+
+# 1. UNINSTALL HELM RELEASE
+kubectl delete \
+	-f "${REPO_ROOT}/test/fixtures/cert-manager-resources.yaml"
+
+helm uninstall \
+    --namespace "${NAMESPACE}" \
+    "$RELEASE_NAME"
+
+kubectl delete "namespace/${NAMESPACE}" \
+	--wait
+
+############
+# VERIFY INSTALL, UPGRADE, UNINSTALL WITH STATIC MANIFESTS
+############
+
+# 1. INSTALL THE LATEST PUBLISHED RELEASE WITH STATIC MANIFESTS
+
+echo "Install cert-manager ${LATEST_RELEASE} using static manifests.."
+kubectl apply \
+	-f "https://github.com/jetstack/cert-manager/releases/download/${LATEST_RELEASE}/cert-manager.yaml" \
+	--wait
+
+kubectl wait \
+	--for=condition=available \
+	--timeout=180s deployment/cert-manager-webhook \
+	--namespace "${NAMESPACE}"
+
+# TODO: This is a dumb sleep that must be removed once we have a check for webhook to be ready.
+echo "Sleep for 30s whilst waiting for cert-manager installation to be ready..."
+sleep 30
+
+# Create a cert-manager issuer and cert
+kubectl apply -f "${REPO_ROOT}/test/fixtures/cert-manager-resources.yaml" \
+	--selector=test=first
+
+# Ensure cert becomes ready
+ensure_cert_ready "test1"
+
+# 2. VERIFY UPGRADE TO THE LATEST BUILD FROM MASTER
+
+echo "Install cert-manager ${CURRENT_VERSION} using static manifests.."
+
+# Build the static manifests
+bazel build //deploy/manifests
+
+# TODO: refactor this functionality here and in
+# devel/addon/certmanager/install.sh so it can be reused.
+# Use the current timestamp as the APP_VERSION so a rolling update will be
+# triggered on every call to this script.
+export APP_VERSION="$(date +"%s")"
+# Build cert-manager images.
+bazel run --stamp=true \
+ --platforms=@io_bazel_rules_go//go/toolchain:linux_amd64 \
+ "//devel/addon/certmanager:bundle"
```

Comment:
Since we've always installed a stable version in the preceding step,
why is it necessary to have a unique APP_VERSION each time?

Reply:
This is for local iterative upgrade testing- the tag with which Bazel tags images will not change if uncommitted changes are made to the codebase so I think that if:
1) someone creates a kind cluster
2) they run the upgrade script 
3) they make some local changes
4) they run the upgrade test again
Without the APP_VERSION tags being different in steps 2 and 4, step 4 would use the same cert-manager images that were loaded to the kind cluster in step 2.
Correct me if I'm wrong- I got this from https://github.com/jetstack/cert-manager/blob/master/devel/addon/certmanager/install.sh#L36

Reply:
I see. I didn't realise that this is also what the `install.sh` script does.
But I'm not convinced that it's needed for the upgrade test, there or here, because if the cert-manager code changes, bazel will build new Docker images with different sha checksums, and then `kind load docker-image` will compare the sha of the local image with the sha of the current image in cluster, and re-upload it because it changed.

In the `install.sh` script, it's necessary to change the tag anyway to trigger the deployment controller into starting a rolling upgrade, but in this upgrade scenario, the image tag will always be different than the LATEST_RELEASE tag.

On the other hand it's probably good to have consistent tags in the static manifest and the helm upgrade tests.
So let's leave it if you agree.

But maybe update the explanatory comment, if you agree that rolling update would always happen in the upgrade test.

Reply:
Thanks, that makes sense. I've updated the comment